### ISSUE TITLE
Boxy Box Pusher

### ISSUE THREAD

**Issue by mistafixxa069** [GDevelopApp/GDevelop-examples#242]
A top-down, grid-based, box pushing game like Sokoban.

## Describe the Example

This is a small Sokoban-like game with several basic systems like:
   -A basic main menu
   -A simple movement system
   -A basic level system
   -Basic box pushing system
   -And a system to check on how many "holes" have been filled
It can be used to make games like A Monster's Expedition, Baba Is You, Stephen's Sausage Roll, etc.
You can only push one box at a time, making it impossible to push boxes with other boxes.

## Checklist

- [x] I've followed all of [the best practices](http://wiki.compilgames.net/doku.php/gdevelop5/community/guide-for-submitting-an-example).
- [x] My game has a proper name in the game properties. 
- [x] My game package name behind with `com.example.`.
- [x] My game has all events unfolded.
- [x] I've added myself as the author in the game properties.
- [x] I've included a file called "README.md" with a description in proper English, explaining what this example is doing.
- [x] I confirm that this game and all of its resources can be integrated to this GitHub repository, distributed and MIT licensed.
- [x] I've cleaned unused resources.

## Game folder

[Boxy Box Pusher.zip](https://github.com/GDevelopApp/GDevelop-examples/files/8452570/Boxy.Box.Pusher.zip)

**Comment by Midhil457** [comment]
Hi, great game!
 I didn't see any issues with the example except, that the readme's short description can be the project's description. Though you can keep the rest. And remove the first title

**Comment by mistafixxa069** [comment]
Hello Leo_Red

May I ask the specifics of what you mean by removing the first title?

Sincerely,
Mistafixxa

**Comment by D8H** [comment]
Thank you for submitting this example. This is a very good start. The events as well organized.

I have some suggestions:
* The history could be a behavior and use an array variable instead of objects. This will avoid to have events for each objects and make the events easier to read.
* The collision system could also be made a behavior with actions like "Character can move 'left' according to Box and Wall". It will make the logic as easy as reading English.
* Custom points can be used to check the collision instead of relying on temporary objects. It's more performant and can be used in extensions.
* Object names must start with uppercase.

**Comment by mistafixxa069** [comment]
Actually, I want you to have this on hold as I have just developed a
simpler and easier example template that might be more efficient and
simpler.

Thank you reviewing this example...

**Comment by mistafixxa069** [comment]
I have updated the example

**Comment by D8H** [comment]
Good job on simplifying the events.

Some collision checks are done every frame. It doesn't affect performance much because Sokoban maps are not huge, but, as it's an example, it's still important to show optimized events.

Currently, the events do this:
- Check if there are any obstacle 2 squares from the player
- Check if there is a box 1 square from the player
- Check in which direction the player is going

It should be done the other way around because checking the player input is really fast and if players don't move left, there is no need to check if they can push something this way.

* There are unnecessary `pick all` conditions.
* Any reason why tween is no longer used? Did you face some limitations with it?
* The goal check can probably be done only at the end of a box push or at the end of a player movement.
* Did you give up the rewind?
* The camera following extension could be used to avoid the big formula.

GridBasedMovement extension
* There should not be "?" in condition names and sentences.
* Condition names (in the sidebar) can start with "Is" when it fit which is the case here
* Avoid condition with negative phrasing `NotMoving` should probably be `IsStopped`.
* Why is the speed multiplied by 4? What is the unit of length of the speed? It should probably be square per second².

**Comment by mistafixxa069** [comment]
Got it, thanks!

**Comment by Midhil457** [comment]
The example can't use multiple file saving option (cuz the system doesn't support)

**Comment by mistafixxa069** [comment]
wait, what do you mean? DM me pls

**Comment by Midhil457** [comment]
Oh, my bad. Didn't realise they were tilesets 😅

**Comment by Bouh** [comment]
Sorry but the example doesn't works see in video:

https://user-images.githubusercontent.com/1670670/167639385-8d87f871-d09d-42eb-b956-0e34be743a54.mp4

**Comment by mistafixxa069** [comment]
huh, will try to fix the issue tommorow

**Comment by mistafixxa069** [comment]
Fixed it

**Comment by Bouh** [comment]
You didn't check what I've done in [my zip file here](https://github.com/GDevelopApp/GDevelop-examples/issues/242#issuecomment-1122561417)
I optimized a lot the resources, the level, and the layers + the zorder, and lot of stuffs.

**Comment by mistafixxa069** [comment]
The object history extension was supposed to be my solution to the undo feature the original had. Also, did I not remove the text? Huh, that's an extra object

**Comment by D8H** [comment]
I've made some changes on top of Bouh version:
- Fix Player movement and animation
- Reorganize the events in groups
- Use Tweens to simplify the player animation
- Make collision check only once when the player chooses a move for efficiency
- Use a small dedicated extension for box pushing
- Remove the camera scrolling because it makes me car ill

https://www.dropbox.com/s/xu9t303bkue61jd/sokoban_REVIEW.zip?dl=1

You said that levels are from Boxxle. This game is probably under copyrights. Levels must be copied with respect of licenses or done from scratch.

**Comment by mistafixxa069** [comment]
Got it.  Will design new levels

**Comment by D8H** [comment]
Hello @mistafixxa069
Any news about this example? It would be a very good example even with only 2 or 3 levels. Maybe there are some open-source level packs that could be used.
